### PR TITLE
Affine, Rotation and Translation transforms

### PR DESCRIPTION
**Pull request by elkins0**
Code and test files to implement geoOps.TrAffine, geoOps.TrRotation and geoOps.TrTranslation to support the TrProjection algorithm AFF, ROT and TRANS types.
Three of the items on the #24 list.

### CODE REVIEW

**Comment by elkins0** [comment]
Using `geoOps._helper.trBuildMatrix` `geoOps.TrProjection.updatePosition` could be simplified to:
```javascript
geoOps.TrProjection.updatePosition = function(el) {
    geoOps._helper.trBuildMatrix(el, function(offset) {
        var a = csgeo.csnames[el.args[0 + offset]].homog,
            b = csgeo.csnames[el.args[2 + offset]].homog,
            c = csgeo.csnames[el.args[4 + offset]].homog,
            d = csgeo.csnames[el.args[6 + offset]].homog;
        return eval_helper.basismap(a, b, c, d);
    });
};
```

**Comment by elkins0** [comment]
Well there you have it, all three parts of the code as separate commits, each with its own test file..

**Comment by gagern** [comment]
Since #144 including ba3cc2a321b95c4869b0ac21dcf755dda661678b got merged, geometric operations need to have a signature describing its allowed input elements. Can you add these please?

(I haven't really looked at the code yet. Hope to do so soon.)

**Comment by elkins0** [comment]
I tried to add these lines today:
```javascript
geoOps.TrAffine.signature = ["P", "P", "P", "P", "P", "P"];
geoOps.TrRotation.signature = ["P", "P", "P", "P"];
geoOps.TrTranslation.signature = ["P", "P"];
```
But I keep getting "Secure Connection Failed". I'll try again tomorrow.

**Comment by elkins0** [comment]
OK --- I've added these lines right after each `geoOps.Tr*.kind = "Tr";` related line:
```javascript
geoOps.TrAffine.signature = ["P", "P", "P", "P", "P", "P"];
geoOps.TrRotation.signature = ["P", "P", "P", "P"];
geoOps.TrTranslation.signature = ["P", "P"];
```

**Comment by elkins0** [comment]
For comparison, here's what `geoOps.TrTranslation.updatePosition` would look like without using `eval_helper.basismap` and `List.adjoint3`:
```javascript
geoOps.TrTranslation.updatePosition = function(el) {
    /*
        Build this matrix when a is [aX, aY, aZ] and  b is [bX, bY, bZ]:

            ⎛aZ*bZ   0    aZ*bX-bZ*aX⎞
        m = ⎜  0   aZ*bZ  aZ*bY-bZ*aY⎟
            ⎝  0     0       aZ*bZ   ⎠
    */
    var a = csgeo.csnames[el.args[0]].homog,
        b = csgeo.csnames[el.args[1]].homog,
        aZ = a.value[2],
        bZ = b.value[2],
        n = CSNumber.mult(aZ, bZ),
        d = List.sub(List.scalmult(aZ, b), List.scalmult(bZ, a)).value,
        zero = CSNumber.zero,
        m = List.turnIntoCSList([
            List.turnIntoCSList([n, zero, d[0]]),
            List.turnIntoCSList([zero, n, d[1]]),
            List.turnIntoCSList([zero, zero, n])
        ]);
    m = List.normalizeMax(m);
    el.matrix = m;
    // Transpose using already normalized values, negate diagonal values
    // Matrix may end up scaled by -1 if n was the max value
    n = CSNumber.neg(m.value[0].value[0]);
    m = List.turnIntoCSList([
        List.turnIntoCSList([n, zero, zero]),
        List.turnIntoCSList([zero, n, zero]),
        List.turnIntoCSList([m.value[0].value[2], m.value[1].value[2], n])
    ]);
    el.dualMatrix = m;
};
```
That's just about as minimal as you can get.

**Comment by gagern** [comment]
My first intuition was that I like this `trBuildMatrix` helper you introduced, since it makes the individual implementations pretty lean. My second thought was that since all you do is call to `basismap` eventually, you might as well provide a helper which does map 4 to 4 and feed that with the right arguments. But then I realized that coming up with the right arguments is a large portion of what makes this code big. So by using the single step function, you only have half as much of that code. So I'm back to liking the approach. :smiley:

I just got some nice idea from this. Instead of writing

    matrix = m1 * adjoint(m0)
    dualMatrix = transpose(adjoint(matrix))

we could try the following instead:

    matrix = m1 * adjoint(m0)
    dualMatrix = transpose(m0 * adjoint(m1))

which does make this setup more symmetric. As a result, we might end up with situations where the dualMatrix is still well defined (albeit not of full rank) even though the primal matrix is all zero. And with your setup, this modification would happen in just a single location.

I think the existing projective transformation implementation should be included in that setup as well. And it should be written to use basismap, too. Any improvements we might want should go into basismap itself.

For the affine projection, I have a different suggestion to make: think about this dually: you have 3 points *A*, *B*, *C*, which define three lines *AB*, *BC*, *AC*. Take these three and map them to their images, and take the line at infinity as fixed. Now you can do the same transformation in the dual world, and only have to swap primal and dual matrix in the end. The benefit is that you'd only have 3 cross products in each step, instead of 7.

I'm still trying to evaluate this whole thing algebraically, to check for common factors and compare implementations in this regard. But this is putting a heavy burden on my CAS, so I don't know when I'll be able to report results.

**Comment by gagern** [comment]
> I'm still trying to evaluate this whole thing algebraically, to check for common factors and compare implementations in this regard. But this is putting a heavy burden on my CAS, so I don't know when I'll be able to report results.

Found out some other task was hogging my CPU.

* The simplification with `dualMatrix = transpose(m0 * adjoint(m1))` is essential in preventing the dual matrix from being far more complicated than the primal one. So I only tested with this.
* With this simplification, the existing `TrProjection` has no common factor. Good to know.
* The `TrRotation` is perfect, too.
* The `TrAffine` still has room for improvement. Even with my simplification, the common factor only dropped from degree 21 to degree 12. It now reads like this:

        (-a2z*b2y*c2x + a2y*b2z*c2x + a2z*b2x*c2y - a2x*b2z*c2y - a2y*b2x*c2z + a2x*b2y*c2z) *
        (a1z*b1y*c1x - a1y*b1z*c1x - a1z*b1x*c1y + a1x*b1z*c1y + a1y*b1x*c1z - a1x*b1y*c1z)^3

  So we have some trouble if three of the defining points are collinear, even though at least one of the two matrices should be well-defined in this setup, too. I don't have an elegant solution for this yet.
* `TrTranslation` has a common factor of `a1z^3*a2z`. Even if you change the implementation to essentially have the oneStep function return the matrix `[[z, 0, x], [0, z, y], [0, 0, z]]` that still leaves you with a `a1z` factor. So I have a tendency to stick to the implementation you pasted above.

@elkins0, would you be willing to [configure me as a collaborator](https://github.com/CindyJS/CindyJS/settings/collaboration) with write permissions for your fork of the project? Then I can commit my suggestions (simpler dual, integrate existing `TrProjection`, simpler affine) into this branch here, so we have these changes here in the pull request for further review.
Would you be willing to grant me

**Comment by gagern** [comment]
I also factorized the common factor of the affine transformation as implemented here in 90db05a9ab9b89c3681b21392a8d63c3ff3faa90, and also in [`evaluator.map$6`](https://github.com/CindyJS/CindyJS/blob/041a9d77d6ce48f88d3630aed9caffa0e0331cb7/src/js/libcs/Operators.js#L3293-L3294) and in Cinderella for Java. It's this:

     c1z * b1z * a1z *
     (-a2z*b2y*c2x + a2y*b2z*c2x + a2z*b2x*c2y - a2x*b2z*c2y - a2y*b2x*c2z + a2x*b2y*c2z)^2 *
     (a1z*b1y*c1x - a1y*b1z*c1x - a1z*b1x*c1y + a1x*b1z*c1y + a1y*b1x*c1z - a1x*b1y*c1z)^4

So one of the things we gain by switching to my implementation (https://github.com/gagern/CindyJS/commit/c5783f1bb9ea675042c16768d8a3ddaea5fa6b24) is that it can still give useful (although singular i.e. non-invertible) results if one of the preimage points is at infinity. Furthermore, the exponents are different, so we might gain some numeric stability using my approach.

The ideal affine transformation matrix, without any leading factors, would be

```
[[a1z*a2z*b1y*b2z*c1z*c2x - a1y*a2z*b1z*b2z*c1z*c2x - a1z*a2z*b1z*b2x*c1y*c2z +  a1z*a2x*b1z*b2z*c1y*c2z + a1y*a2z*b1z*b2x*c1z*c2z - a1z*a2x*b1y*b2z*c1z*c2z,
  -a1z*a2z*b1x*b2z*c1z*c2x + a1x*a2z*b1z*b2z*c1z*c2x + a1z*a2z*b1z*b2x*c1x*c2z -a1z*a2x*b1z*b2z*c1x*c2z - a1x*a2z*b1z*b2x*c1z*c2z + a1z*a2x*b1x*b2z*c1z*c2z,
  a1y*a2z*b1x*b2z*c1z*c2x - a1x*a2z*b1y*b2z*c1z*c2x - a1y*a2z*b1z*b2x*c1x*c2z + a1z*a2x*b1y*b2z*c1x*c2z + a1x*a2z*b1z*b2x*c1y*c2z - a1z*a2x*b1x*b2z*c1y*c2z],
 [a1z*a2z*b1y*b2z*c1z*c2y - a1y*a2z*b1z*b2z*c1z*c2y - a1z*a2z*b1z*b2y*c1y*c2z + a1z*a2y*b1z*b2z*c1y*c2z + a1y*a2z*b1z*b2y*c1z*c2z - a1z*a2y*b1y*b2z*c1z*c2z,
  -a1z*a2z*b1x*b2z*c1z*c2y + a1x*a2z*b1z*b2z*c1z*c2y + a1z*a2z*b1z*b2y*c1x*c2z - a1z*a2y*b1z*b2z*c1x*c2z - a1x*a2z*b1z*b2y*c1z*c2z + a1z*a2y*b1x*b2z*c1z*c2z,
  a1y*a2z*b1x*b2z*c1z*c2y - a1x*a2z*b1y*b2z*c1z*c2y - a1y*a2z*b1z*b2y*c1x*c2z + a1z*a2y*b1y*b2z*c1x*c2z + a1x*a2z*b1z*b2y*c1y*c2z - a1z*a2y*b1x*b2z*c1y*c2z],
 [0,
  0,
  a1z*a2z*b1y*b2z*c1x*c2z - a1y*a2z*b1z*b2z*c1x*c2z - a1z*a2z*b1x*b2z*c1y*c2z + a1x*a2z*b1z*b2z*c1y*c2z + a1y*a2z*b1x*b2z*c1z*c2z - a1x*a2z*b1y*b2z*c1z*c2z]]

I don't think we should implement this manually, but I would welcome any approach which comes up with this result in a more elegant way.

**Comment by elkins0** [comment]
@gagern: I've added you as a collaborator using that link. There was nothing more to choose as far as I could see having to do with permissions so you'll have to let me know if there's anything more I need to do...

**Comment by gagern** [comment]
@elkins0: Thanks, I could commit. I could even rebase to resolve a conflict with recent code related to Möbius transformations.

@strobelm, since this now contains code from me as well, will you have a final look at things?

**Comment by elkins0** [comment]
So I took a look at the ideal affine transformation matrix. I decided to group triple multiples of symbols in the following way creating new symbols for each group. These are shown in Mathematica replacement rule form:
```mathematica
{a1x b1z c1z -> sa1x, a1y b1z c1z -> sa1y, a2x b2z c2z -> sa2x,  a2y b2z c2z -> sa2y, 
 a1z b1x c1z -> sb1x, a1z b1y c1z -> sb1y, a2z b2x c2z -> sb2x, a2z b2y c2z -> sb2y,
 a1z b1z c1x -> sc1x, a1z b1z c1y -> sc1y, a2z b2z c2x -> sc2x,  a2z b2z c2y -> sc2y, 
 a1z b1y c1x -> bycx1t, a1y b1z c1x -> aycx1t, 
 a1z b1x c1y -> bxcy1t, a1x b1z c1y -> axcy1t, 
 a1y b1x c1z -> aybx1t, a1x b1y c1z -> axby1t, a2z b2z c2z -> abc2z}
```
I then proceeded to extract what appeared to be dot product groupings and came up with the following simplification as the dot product of two matrices, again shown in Mathematica form:
```mathematica
Expand[Transpose[
    {{sc1y - sb1y, sa1y - sc1y, sb1y - sa1y},
     {sb1x - sc1x, sc1x - sa1x, sa1x - sb1x},
     {bycx1t - bxcy1t, axcy1t - aycx1t, aybx1t - axby1t}}.Transpose[
    {{sa2x, sb2x, sc2x}, {sa2y, sb2y, sc2y}, {abc2z, abc2z, abc2z}}]]]
```
Here's a Mathematica file [AffineTransformSimplified.nb](https://github.com/elkins0/CindyJS/blob/patch-3/tests/AffineTransformSimplified.nb) showing the application of the rules to the ideal affine transformation matrix and showing the equivalence of the simplified result.

**Comment by gagern** [comment]
I think I got a good geometric handle on the AffineTransform. Will propose that soon.

**Comment by elkins0** [comment]
Sorry. Guess the build doesn't ignore *.nb files :smile: -- I just didn't want to make a huge comment entry. Anyway, I choose nine of the symbol names (those strating with s) to mean scaled. The six more are just suffixed by 1t. Note how much computation goes into creating those two zeros in the matrix!

**Comment by gagern** [comment]
@elkins0 Usually [gist](https://gist.github.com/) is better suited for this kind of “extended comment”. Mind if I remove that change from this branch in a subsequent (forced) push?

**Comment by elkins0** [comment]
Thanks for the tip. Yes toss it. [AffineTransformSimplified.nb](https://gist.github.com/elkins0/f5a98a5ae98b8a8c7571) is at the new location. I replaced the inner transform with its result while I was at it. Makes more sense, the transform result:
```mathematica
{{sa2x, sa2y, abc2z}, {sb2x, sb2y, abc2z}, {sc2x, sc2y, abc2z}}
```
is just the list of the image points' scaled homogeneous coordinates.

**Comment by gagern** [comment]
[Here](https://github.com/CindyJS/CindyJS/files/65335/TrAffine.pdf) is what I worked out in the meantime. Now I just have to implement this.

**Comment by elkins0** [comment]
And finally, 
```mathematica
{bycx1t - bxcy1t, axcy1t - aycx1t, aybx1t - axby1t}
```
is just
```mathematica
-Cross[{a1x, b1x, c1x}, {a1y, b1y, c1y}] {a1z, b1z, c1z}
```
or
```mathematica
-Cross @@ Most [#] Last[#] &@
 Transpose[{{a1x, a1y, a1z}, {b1x, b1y, b1z}, {c1x, c1y, c1z}}]
```

**Comment by gagern** [comment]
OK, I implemented my approach. Now the primal matrix is non-zero even if the target points are collinear, and the dual matrix is non-zero for collinear source points.

@elkins0 I confess I don't read Mathematica this fluently. Particularly since I don't have a licensed copy available here. Do you think what you computed there is any shorter or easier to understand than what I implemented in [my latest commit](https://github.com/CindyJS/CindyJS/commit/986d62cfc46adb791bd3700010e17a88ee853222)? If so, go ahead and write an implementation for comparison. If not, I'd say we should go with my implementation of the operation, and ask @strobelm and @kranich about what kind of explaining comments they'd like to see included.

In either case, I think we should take your implementation of the translation from https://github.com/CindyJS/CindyJS/pull/148#issuecomment-164997426. Would you include that here?

**Comment by gagern** [comment]
One more thing I'd like to discuss before merging: the thing currently called `TrRotation` is in fact a similarity transformation, not a purely isomentric rotation, right? If so I think it would make sense to rename it appropriately in the Cinderella export, and change the name to `TrSimilarity` here.

**Comment by elkins0** [comment]
Nice work! I like it better than what I came up with. However you no longer need:
```javascript
     var inf = List.linfty;
```
Also `mult(a1z, mult(b2z, c2z))` becomes `multiMult([b2z, c2z, a1z])` and so on in six places.

For comparison, here's what I came up with:
```javascript
    var mult = CSNumber.mult;
    var scalmult = List.scalmult;
    // Scale each point's homogeneous coordinates by the
    // other two point's z-value for each set of three points.
    function scale(ps) {
        var v = ps.value,
            z0 = v[0].value[2],
            z1 = v[1].value[2],
            z2 = v[2].value[2];
        return List.turnIntoCSList([
            scalmult(mult(z1, z2), v[0]),
            scalmult(mult(z0, z2), v[1]),
            scalmult(mult(z0, z1), v[2])
        ]);
    }
    // Get the set of points and the set of thier images
    var ps1 = List.turnIntoCSList([
            csgeo.csnames[el.args[0]].homog,
            csgeo.csnames[el.args[2]].homog,
            csgeo.csnames[el.args[4]].homog
    ]);
    // Get the set of thier images
    var ps2 = List.turnIntoCSList([
            csgeo.csnames[el.args[1]].homog,
            csgeo.csnames[el.args[3]].homog,
            csgeo.csnames[el.args[5]].homog
    ]);
    var tsm1v = List.transpose(scale(ps1)).value;
    // Compute row 3 of first matrix
    var t1r3v = List.transpose(ps1).value;
    var cplv = List.neg(List.cross(t1r3v[0], t1r3v[1])).value;
    var t1r3v2 = t1r3v[2].value;
    var m = List.transpose(List.productMM(
        List.turnIntoCSList([
            List.cross(tsm1v[1], List.realVector([-1, -1, -1])),
            List.cross(tsm1v[0], List.realVector([1, 1, 1])),
            List.turnIntoCSList([mult(cplv[0], t1r3v2[0]), mult(cplv[1], t1r3v2[1]), mult(cplv[2], t1r3v2[2])])
        ]),
        scale(ps2)));
    m = List.normalizeMax(m);
    el.matrix = m;
    m = List.transpose(List.adjoint3(m));
    m = List.normalizeMax(m);
    el.dualMatrix = m;
```

**Comment by elkins0** [comment]
OK, I've updated geoOps.TrTranslation.updatePosition. You may also want to update the inline function within `geoOps.TrProjection.updatePosition` to return `eval_helper.basismap(a, b, c, d);`.

**Comment by elkins0** [comment]
I had a very symmetric moment. I took the best of your very fine work and the best of mine :smile: and this is the result is glorious *I think*.

**Comment by elkins0** [comment]
*Nothing* like having the build rain on your parade at the last possible moment. I'm out of time for now...

**Comment by elkins0** [comment]
After collaboration report...

So I looked at my bottom up solution and @garern's top down solution and wondered:
How are they different?
* We both have defined the result matrix in similar ways.
* Both have used the dot product of two 3&times;3 matrices.
* I did't bother doing anything special to get the dualMatrix.
* Neither of ours was completely symmetric.

The following is what intrigued me about the two 3&times;3 matrices.

Left versus right (matrix then dualMatrix):
```javascript
  t(mat([a2, b2, c2])) and ... sm(mult(a1z, mult(b2z, c2z)), cc(b1, c1)) ...
  t(mat([cc(b2, c2), cc(c2, a2), cc(a2, b2)])) and ... sm(mult(a2z, mult(b1z, c1z)), a1) ...
```
and for comparison one of mine was `scale(ps1)` where ps2 was `mat([a2, b2, c2])`.

Now @garern's implementation has all of the scaling (as it were) unevenly distributed and not grouped with the point sets.
What if we play fast and loose and re-associate things a wee bit, like so, using the pieces inexactly from above, grouping 1's and 2's:
```javascript
  t(sm(mult(b2z, c2z), ...) mat([a2, b2, c2])) and ... sm(a1z, cc(b1, c1)) ...
  t(sm(mult(a2z, ) ...) mat([cc(b2, c2), cc(c2, a2), cc(a2, b2)])) and ... sm(mult(b1z, c1z), a1) ...
```
This gives a symmetric pattern which is produced by the two functions `scale` and `crossScale` in the merged code.

I checked the resulting matrix symbolically using *Mathematica* and compared it to @gagern's CAS result. As it turns out it is scaled by minus one:
```mathematica
Block[{
    a1={a1x,a1y,a1z},a2={a2x,a2y,a2z},
    b1={b1x,b1y,b1z},b2={b2x,b2y,b2z},
    c1={c1x,c1y,c1z},c2={c2x,c2y,c2z},ps1,ps2,
    scale={
        #2[[3]]#3[[3]]#1,
        #1[[3]]#3[[3]]#2,
        #1[[3]]#2[[3]]#3}&,
    crossScale={
        #1[[3]]Cross[#2,#3],
        #2[[3]]Cross[#3,#1],
        #3[[3]]Cross[#1,#2]}&},
ps1={a1,b1,c1};
ps2={a2,b2,c2};
Expand[-Transpose[scale@@ps2].crossScale@@ps1]]==matrix1
```
Note: [AffineTransformSimplified.nb](https://gist.github.com/elkins0/f5a98a5ae98b8a8c7571) defines matrix1.

**Comment by elkins0** [comment]
I determined that `t(mat([cc(b2, c2), cc(c2, a2), cc(a2, b2)]))` is really just `List.adjoint3(mat([a2, b2, c2]))` in @gagern's original coding. So I re-worked `geoOps.TrAffine.updatePosition` yet again since I didn't need two different scaling functions to make the code symmetrical. I put the ps1 and ps2 z-values into a list using List.column. The result is still just two lines with the setup out of the way:
```javascript
    el.matrix = nm(mm(t(ps2), scl(z1, z2, t(adj(ps1)))));
    el.dualMatrix = nm(mm(adj(ps2), scl(z2, z1, ps1)));
```

**Comment by elkins0** [comment]
I used *Mathematica* to determine that the adjoint and the negative of the transpose of @gagern's matrix of cross products were equivalent:
```mathematica
Block[{a = {ax, ay, az}, b = {bx, by, bz}, c = {cx, cy, cz}, 
  adjoint = Map[Reverse, Minors[Transpose[#], Length[#] - 1], {0, 1}]*
     Table[(-1)^(i + j), {i, Length[#]}, {j, Length[#]}] &, 
  tCross = MapThread[Cross, {RotateRight@#, RotateLeft@#}] &}, 
 adjoint[{a, b, c}] == -Transpose[tCross[{a, b, c}]]]
```
Each products the following matrix:
```mathematica
{{-bz cy + by cz,  az cy - ay cz, -az by + ay bz},
 { bz cx - bx cz, -az cx + ax cz,  az bx - ax bz},
 {-by cx + bx cy,  ay cx - ax cy, -ay bx + ax by}}
```
Also, using *Mathematica*, I determined that I still get the matrix from @gagern's CAS result:
```mathematica
Block[{a1 = {a1x, a1y, a1z}, a2 = {a2x, a2y, a2z}, 
   b1 = {b1x, b1y, b1z}, b2 = {b2x, b2y, b2z}, c1 = {c1x, c1y, c1z}, 
   c2 = {c2x, c2y, c2z}, ps1, ps2, z1, z2, 
   adj = Map[Reverse, Minors[Transpose[#], Length[#] - 1], {0, 1}]*
      Table[(-1)^(i + j), {i, Length[#]}, {j, Length[#]}] &, 
   scl = Times[#1, Sequence[Through[Times[RotateRight, RotateLeft][#2]]], #3] &}, 
  ps1 = {a1, b1, c1};
  ps2 = {a2, b2, c2};
  z1 = Part[ps1, All, 3];
  z2 = Part[ps2, All, 3];
  Expand[-Transpose[ps2].scl[z1, z2, Transpose[adj[ps1]]]]] == matrix1
```
Note: [AffineTransformSimplified.nb](https://gist.github.com/elkins0/f5a98a5ae98b8a8c7571) defines matrix1.

**Comment by elkins0** [comment]
Found a way to get a list of z-values without having to resort to using `List.column` by getting the last row from the transpose of the point sets. Switched `t(adj(ps1))` to `adj(t(ps1))` to avoid an unnecessary `List.transpose` call to do so. Also collected duplicate `List.mult` calls into a vector to allow their reuse. So now calls to `List.multiMult` are no longer needed and the `scl` function is now in-lined due to its simplicity. That's it for any optimizations I can see to the `geoOps.TrAffine.updatePosition` code.

**Comment by strobelm** [comment]
Hi, since there were still changes in the code lately @elkins0 and @gagern do you think it's stable enough to review and possibly merge this?

**Comment by gagern** [comment]
@strobelm I added some things I had in my TODO queue, now things should be good from my side. Recent changes by @elkins0 were regarding the notation for the `TrAffine` computation. I think that should have converged to something close to optimal by now, too. I'd be glad to see this merged.

**Comment by elkins0** [comment]
Yes, please review @strobelm. I'm done tinkering with the code unless you or @gagern have any other recommendations.

File: examples/TrTranslation.html [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+	defaultAppearance: { fontFamily: "sans-serif", lineSize: 1, pointSize: 5.0 }, 
+	angleUnit: "°", 
+	geometry: [ 
+		{ name: "P1", type: "Free", pos: [ 4.0, -4.0, -0.6666666666666666 ], color: [ 0.0, 0.0, 0.0 ], labeled: true, size: 3.0 }, 
+		{ name: "P1i", type: "Free", pos: [ 0.0, -4.0, -1.3333333333333333 ], color: [ 0.5019608, 0.5019608, 0.5019608 ], labeled: true, size: 3.0, printname: "P1'" }, 
+		{ name: "Tr0", type: "TrTranslation", color: [ 0.0, 0.0, 1.0 ], args: [ "P1", "P1i" ] }, 
+		{ name: "O", type: "Free", pos: [ 4.0, 3.3333333333333335, 0.6666666666666666 ], color: [ 1.0, 1.0, 1.0 ], labeled: true }, 
+		{ name: "C0", type: "CircleByRadius", color: [ 1.0, 1.0, 0.0 ], radius: 2.999999999999999, args: [ "O" ], printname: "$C_{0}$" }, 
+		{ name: "X", type: "PointOnCircle", pos: [ 4.0, { r: 2.2077737946188507, i: -1.3419095945542774E-16 }, { r: 0.4444531955041505, i: 1.6304775351968118E-19 } ], color: [ 0.0, 0.0, 1.0 ], args: [ "C0" ], labeled: true, size: 3.0 }, 
+		{ name: "C1", type: "CircleByRadius", color: [ 1.0, 1.0, 0.0 ], radius: 0.041785545347643845, args: [ "X" ], printname: "$C_{1}$" },
```
Comment (strobelm) [inline examples/TrTranslation.html]: Any reason for that circle in the example? Sometimes it produces ghost lines.

**Comment by strobelm** [comment]
Quite an effort you have put into that! The code looks good although i didn't understand the affine transformations without your comments.  Examples are looking good and behave like in Cinderella.

I will add some comments, squash the commits and merge it into master.

**Comment by strobelm** [comment]
And the code is in the master branch, thanks again @elkins0 and @gagern!

File: examples/TrTranslation.html (hunk 1 above)
Comment (elkins0) [inline examples/TrTranslation.html]: Oh&mdash;that is superfluous&mdash;that relates to what I reported as issue #139. I didn't realize I'd left it in the example. Referring to:
```javascript
{ name: "C1", type: "CircleByRadius", color: [ 1.0, 1.0, 0.0 ], radius: 0.041785545347643845, args: [ "X" ], printname: "$C_{1}$" }, 
```
[Zero radius circles sometimes drawn as a line](https://github.com/CindyJS/CindyJS/issues/139)&mdash;in this case the radius is tiny but not zero.

A point [x y 1] can be represented as a circle with the following matrix:
```
⎛ 1   0  -x ⎞
⎜ 0   1  -y ⎟
⎝-x  -y   0 ⎠
```
Whereas this circle matrix:
```
⎛ 0   0  a/2⎞
⎜ 0   0  b/2⎟
⎝a/2 b/2  1 ⎠
```
factors into two lines with homogeneuous coordinates of [a b 1] and [0 0 1].

**Comment by gagern** [comment]
Thanks for the review, @strobelm!

**Pull request closed without merging**